clustermesh/clustermesh.go

```diff
@@ -1885,3 +1941,199 @@ func DisableWithHelm(ctx context.Context, k8sClient *k8s.Client, params Paramete
 	_, err = helm.Upgrade(ctx, k8sClient.RESTClientGetter, upgradeParams)
 	return err
 }
+
+func getRelease(kc *k8s.Client, namespace string) (*release.Release, error) {
+	client := kc.RESTClientGetter
+	release, err := helm.GetCurrentRelease(client, namespace, defaults.HelmReleaseName)
+	if err != nil {
+		return nil, err
+	}
+	return release, nil
+}
+
+func (k *K8sClusterMesh) validateCAMatch(aiLocal, aiRemote *accessInformation) error {
+	caLocal := aiLocal.CA
+	block, _ := pem.Decode(caLocal)
+	if block == nil {
+		panic("failed to parse certificate PEM")
+	}
+	localCert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return err
+	}
+
+	caRemote := aiRemote.CA
+	block, _ = pem.Decode(caRemote)
+	if block == nil {
+		panic("failed to parse certificate PEM")
+	}
+	remoteCert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return err
+	}
+
+	// Compare the x509 key identifier of each certificate
+	if !bytes.Equal(localCert.SubjectKeyId, remoteCert.SubjectKeyId) {
+		return fmt.Errorf("Local and Remote cluster cilium-ca certificate keys do not match")
+	}
+	return nil
+}
+
+// ConnectWithHelm enables clustermesh using a Helm Upgrade
+// action. Certificates are generated via the Helm chart's cronJob
+// (certgen) mode. As with classic mode, only autodetected IP-based
+// clustermesh-apiserver Service endpoints are currently supported.
+func (k *K8sClusterMesh) ConnectWithHelm(ctx context.Context) error {
+	localRelease, err := getRelease(k.client.(*k8s.Client), k.params.Namespace)
+	if err != nil {
+		k.Log("❌ Unable to find Helm release for the target cluster")
+		return err
+	}
+	version := localRelease.Chart.AppVersion()
+	semv, err := utils.ParseCiliumVersion(version)
+	if err != nil {
+		return fmt.Errorf("failed to parse Cilium version: %w", err)
+	}
+
+	v := fmt.Sprintf("%d.%d.%d", semv.Major, semv.Minor, semv.Patch)
+	k.Log("✅ Detected Helm release with Cilium version %s", v)
+
+	if v < "1.14.0" {
+		// Helm-based clustermesh enable is only supported on Cilium v1.14+ due to a lack of support in earlier versions
+		// for autoconfigured certificates (tls.{crt,key}) for cluster members when running in certgen (cronJob) PKI
+		// mode
+		k.Log("⚠️ Cilium Version is less than 1.14.0. Continuing in classic mode.")
+		return k.Connect(ctx)
+	}
+
+	localClient, remoteClient, err := k.getClientsForConnect()
+	if err != nil {
+		return err
+	}
+
+	aiLocal, aiRemote, err := k.getAccessInfoForConnect(ctx, localClient, remoteClient)
+	if err != nil {
+		return err
+	}
+
+	err = k.validateInfoForConnect(aiLocal, aiRemote)
+	if err != nil {
+		return err
+	}
+
+	// Validate that CA certificates match between the two clusters
+	err = k.validateCAMatch(aiLocal, aiRemote)
+	if err != nil {
+		return err
+	}
```

Comment:
In the default case (using `cronJob` mode), we would have the `caCert` for the `tls` config but not the `cert` or `key`. Does this config now support passing only the `caCert` for each cluster?

Reply:
I think that the `cert` and `key` info are already retrieved through `extractAccessInformation` (unless there is a secret name mismatch, which is 100% possible). That key/cert pair in the secret will be overwritten the subsequent time the certgen job is run (most likely when helm is executed to connect the two clusters), but the old ones will still work. I agree that this is not the best solution, but it might help to smooth the user experience when the clusters are configured with a different CA (which I think is fairly common especially in test setups). WDYT?

Reply:
One of the reasons for defaulting to `crontJob` mode is so that certificate renewals will happen automatically. Setting `tls:` at connect time would diminish that benefit.

This does raise an interesting concern which is that some users may wish to set `address:` at connect time, and the current implementation would overwrite that silently.

I need to think about this more.

Reply:
> One of the reasons for defaulting to crontJob mode is so that certificate renewals will happen automatically. Setting tls: at connect time would diminish that benefit.

I 100% agree. If we decide to go with this, there should be a warning that it is a compatibility mode with many limitations (including at lease hubble issues and lack of certs renewal). 

> This does raise an interesting concern which is that some users may wish to set address: at connect time, and the current implementation would overwrite that silently.

I'm not sure I got what you mean here.

Reply:
> I'm not sure I got what you mean here.

We know that some users are interested in this connect implementation using hostnames (possibly not resolving `ips:` at all, see https://github.com/cilium/cilium-cli/issues/1539). The current implementation of the config merging in this PR, when it constructs the config entry for the remote cluster, does not include any `address:` even if there was an attempt to provide it with `--helm-set` or `--helm-file`.

Reply:
Gotcha. Yeah, I didn't try that possibility since I saw in the comments that `address` was not yet supported.

Reply:
Implemented your initial suggestion of warning the user on CA mismatch and configuring `tls:` https://github.com/cilium/cilium-cli/pull/1628/commits/82eedfceb01fdf0a8a14dc221ca50a17bfd41ff9